File: src/test/java/org/geepawhill/contentment/geometry/BezierTest.java
package org.geepawhill.contentment.geometry;

import static org.assertj.core.api.Assertions.within;
import static org.geepawhill.contentment.test.ContentmentAssertions.assertThat;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class BezierTest
{

	private Point start;
	private Point end;
	private Bezier linear;

	@BeforeEach
	public void before()
	{
		start = new Point(0d, 0d);
		end = new Point(100d, 100d);
		linear = new Bezier(start, end);
	}

	@Test
	public void constructor()
	{
		PointPair line = new PointPair(linear.getStart(), linear.getEnd());
		assertThat(linear.getStart()).isEqualTo(start);
		assertThat(linear.getEnd()).isEqualTo(end);
		assertThat(linear.getHandle1()).isEqualTo(line.along(1d / 3d));
		assertThat(linear.getHandle2()).isEqualTo(line.along(2d / 3d));
	}

	@Test
	public void along()
	{
		assertThat(linear.along(0d)).isEqualTo(start);
		assertThat(linear.along(1d)).isEqualTo(end);
		assertThat(linear.along(.5d)).isGoodEnough(new PointPair(start, end).center(), within(0.1d));
		for (int i = 0; i < 11; i++)
		{
			double fraction = ((double) i) / 10;
			assertThat(linear.along(fraction)).isGoodEnough(new PointPair(0d, 0d, 100d, 100d).along(fraction), within(0.1d));
			assertThat(linear.along(fraction)).isGoodEnough(new PointPair(0d, 0d, 100d, 100d).along(fraction), within(0.1d));
		}
	}
	
	@Test
	public void linearHalf()
	{
		Bezier split = linear.split(.5d);
		assertThat(split.getEnd()).isGoodEnough(new Point(50d, 50d), within(0.1d));
	}

	@Test
	public void linearQuarter()
	{
		Bezier split = linear.split(.25d);
		assertThat(split.getEnd()).isGoodEnough(new Point(25d, 25d), within(0.1d));
	}


}

File: src/test/kotlin/org/geepawhill/contentment/geometry/BezierTest.kt
package org.geepawhill.contentment.geometry

import org.assertj.core.api.Assertions.within
import org.geepawhill.contentment.test.ContentmentAssertions.Companion.assertThat
import org.junit.jupiter.api.Test

class BezierTest {

    private val start = Point(0.0, 0.0)
    private val end = Point(100.0, 100.0)
    private val linear = Bezier(start, end)

    @Test
    fun constructor() {
        val line = PointPair(linear.start, linear.end)
        assertThat(linear.start).isEqualTo(start)
        assertThat(linear.end).isEqualTo(end)
        assertThat(linear.handle1).isEqualTo(line.along(1.0 / 3.0))
        assertThat(linear.handle2).isEqualTo(line.along(2.0 / 3.0))
    }

    @Test
    fun along() {
        assertThat(linear.along(0.0)).isEqualTo(start)
        assertThat(linear.along(1.0)).isEqualTo(end)
        assertThat(linear.along(.5)).isGoodEnough(PointPair(start, end).center(), within(0.1))
        for (i in 0..10) {
            val fraction = i.toDouble() / 10
            assertThat(linear.along(fraction)).isGoodEnough(PointPair(0.0, 0.0, 100.0, 100.0).along(fraction), within(0.1))
            assertThat(linear.along(fraction)).isGoodEnough(PointPair(0.0, 0.0, 100.0, 100.0).along(fraction), within(0.1))
        }
    }

    @Test
    fun linearHalf() {
        val split = linear.split(.5)
        assertThat(split.end).isGoodEnough(Point(50.0, 50.0), within(0.1))
    }

    @Test
    fun linearQuarter() {
        val split = linear.split(.25)
        assertThat(split.end).isGoodEnough(Point(25.0, 25.0), within(0.1))
    }


}
